apps/remote_control/lib/lexical/remote_control/code_intelligence/rename/module.ex

```diff
@@ -0,0 +1,164 @@
+    # For instance, when there are no more submodules after the cursor.
+    # like: `defmodule TopLevel.Mo|dule do`
+    # in most cases, users only want to rename the module itself.
+    #
+    # However, there's an exception when the cursor is in the middle,
+    # such as `Top.Mo|dule.ChildModule`. If we rename it to `Top.Renamed.Child`,
+    # it would be natural to also rename `Module.ChildModule` to `Renamed.Child`.
+    if at_the_middle_of_module?(document, position, range) do
+      entities ++ descendants(entity, local_module_name)
+    else
+      entities
+    end
+  end
+
+  defp at_the_middle_of_module?(document, position, range) do
+    range_text = range_text(range)
+
+    case Ast.surround_context(document, position) do
+      {:ok, %{context: {:alias, alias}}} ->
+        String.length(range_text) < length(alias)
+
+      _ ->
+        false
+    end
+  end
+
+  defp descendants(entity, local_module_name) do
+    entity_string = inspect(entity)
+    prefix = "#{entity_string}."
+
+    prefix
+    |> Store.prefix(type: :module)
+    |> Enum.filter(&(entry_matching?(&1, local_module_name) and has_dots_in_range?(&1)))
+    |> adjust_range(entity)
+  end
+
+  defp exacts(entity, local_module_name) do
+    entity_string = inspect(entity)
+
+    entity_string
+    |> Store.exact(type: :module)
+    |> Enum.filter(&entry_matching?(&1, local_module_name))
+  end
+
+  defp entry_matching?(entry, local_module_name) do
+    entry.range |> range_text() |> String.contains?(local_module_name)
+  end
+
+  defp has_dots_in_range?(entry) do
+    entry.range |> range_text() |> String.contains?(".")
+  end
+
+  defp adjust_range(entries, entity) do
+    for entry <- entries,
+        uri = Document.Path.ensure_uri(entry.path),
+        range_result = resolve_local_module_range(uri, entry.range.start, entity),
+        match?({:ok, _}, range_result) do
+      {_, range} = range_result
+      %{entry | range: range}
+    end
+  end
+
+  defp resolve_local_module_range(uri, position, entity) do
+    with {:ok, _} <- Document.Store.open_temporary(uri),
+         {:ok, document, analysis} <- Document.Store.fetch(uri, :analysis),
+         {:ok, result, range} <- resolve_module(analysis, position) do
+      if result == entity do
+        {:ok, range}
+      else
+        last_part_length = result |> Ast.Module.local_module_name() |> String.length()
+        # Move the cursor to the next part:
+        # `|Parent.Next.Target.Child` -> 'Parent.|Next.Target.Child' -> 'Parent.Next.|Target.Child'
+        character = position.character + last_part_length + 1
+        position = Position.new(document, position.line, character)
+        resolve_local_module_range(uri, position, entity)
+      end
+    else
+      _ ->
+        Logger.error("Failed to find entity range for #{inspect(uri)} at #{inspect(position)}")
+        :error
+    end
+  end
+
+  defp to_edits_by_uri(results, new_name) do
+    Enum.group_by(
```

Comment:
nitpick, the group_by should not be in this function, it should be in the caller. It's  more flexible to deal with an individual element than a group of things. The caller would be
```elixir
|> Enum.group_by(& &Document.Path.ensure_uri(&1.path), &result_to_edits/1))
```